frontend/include/chpl/framework/Context-detail.h

```diff
@@ -282,28 +293,61 @@ class QueryMapResult final : public QueryMapResultBase {
   //  * a default-constructed result
   QueryMapResult(QueryMap<ResultType, ArgTs...>* parentQueryMap,
                  std::tuple<ArgTs...> tupleOfArgs)
-    : QueryMapResultBase(-1, -1, false, parentQueryMap),
+    : QueryMapResultBase(-1, -1, false, {}, parentQueryMap),
       tupleOfArgs(std::move(tupleOfArgs)),
       result() {
   }
   QueryMapResult(RevisionNumber lastChecked,
                  RevisionNumber lastChanged,
                  bool emittedErrors,
+                 std::set<const QueryMapResultBase*> recursionErrors,
                  QueryMap<ResultType, ArgTs...>* parentQueryMap,
                  std::tuple<ArgTs...> tupleOfArgs,
                  ResultType result)
-    : QueryMapResultBase(lastChecked, lastChanged, emittedErrors, parentQueryMap),
+    : QueryMapResultBase(lastChecked, lastChanged, emittedErrors,
+                         std::move(recursionErrors), parentQueryMap),
       tupleOfArgs(std::move(tupleOfArgs)),
       result(std::move(result)) {
   }
   void recompute(Context* context) const override;
   void markUniqueStringsInResult(Context* context) const override;
 };
 
+// ID is only forward declared here, so we cannot return it by value. Thus,
+// return an ID ref to the argument tuple; calling code will need to perform
+// a copy to construct an actual TraceElement.
```

Comment:
I would think that it's possible to `#include "chpl/framework/ID.h"` without including `Context.h` or `Context-detail.h`. It might be nice to avoid the workaround you have here as it has implications for the calling code.